generators/generate_interactions_personalize.py

```diff
@@ -0,0 +1,393 @@
+for products in the same category. This is a hotfix for the logic whereby we generate profiles and probabilistically
+sample product categories according to the sample user profile. Item 4 is necessary for training the discounts
+personalizeation campaign.
+"""
+import json
+import pandas as pd
+import numpy as np
+import time
+import csv
+from pathlib import Path
+import gzip
+import random
+import yaml
+import logging
+from collections import defaultdict
+
+# Keep things deterministic
+RANDOM_SEED = 0
+
+# Without checking the contents, skip making any output files. Since this script is deterministic,
+# this will usually give the correct behaviour.
+SKIP_EXISTING = True
+
+# Where to put the generated data so that it is picked up by stage.sh
+GENERATED_DATA_ROOT = "src/aws-lambda/personalize-pre-create-campaigns/data"
+
+# Interactions will be generated between these dates
+FIRST_TIMESTAMP = 1591803782  # 2020-06-10, 18:43:02
+LAST_TIMESTAMP = 1599579782  # 2020-09-08, 18:43:02
+
+# Users are set up with 3 product categories on their personas. If [0.6, 0.25, 0.15] it means
+# 60% of the time they'll choose a product from the first category, etc.
+CATEGORY_AFFINITY_PROBS = [0.6, 0.25, 0.15]
+
+# After a product, there are this many products within the category that a user is likely to jump on next.
+# The purpose of this is to keep recommendations focused within the category if there are too many products
+# in a category, because at present the user profiles approach samples products from a category.
+PRODUCT_AFFINITY_N = 4
+
+# from 0 to 1. If 0 then products in busy categories get represented less. If 1 then all products same amount.
+NORMALISE_PER_PRODUCT_WEIGHT = 1.0
+
+# With this probability a product interaction will be with the product discounted
+# Here we go the other way - what is the probability that a product that a user is already interacting
+# with is discounted - depending on whether user likes discounts or not
+DISCOUNT_PROBABILITY = 0.2
+DISCOUNT_PROBABILITY_WITH_PREFERENCE = 0.5
+
+IN_PRODUCTS_FILENAME = "src/products/src/products-service/data/products.yaml"
+IN_USERS_FILENAME = "src/users/src/users-service/data/users.json.gz"
+
+PROGRESS_MONITOR_SECONDS_UPDATE = 30
+
+# This is where stage.sh will pick them up from
+out_items_filename = f"{GENERATED_DATA_ROOT}/items.csv"
+out_users_filename = f"{GENERATED_DATA_ROOT}/users.csv"
+out_interactions_filename = f"{GENERATED_DATA_ROOT}/interactions.csv"
+
+# The meaning of the below constants is described in the relevant notebook.
+
+# Minimum number of interactions to generate
+min_interactions = 675000
+# min_interactions = 50000
+
+# Percentages of each event type to generate
+product_added_percent = .08
+cart_viewed_percent = .05
+checkout_started_percent = .02
+order_completed_percent = .01
+
+CATEGORY_AFFINITY_PROBS = np.array(CATEGORY_AFFINITY_PROBS)
+
+
+def generate_data():
```

Comment:
This implementation adds a third algorithm/approach for creating interactions in the project. 

- The original interaction generator in the Personalize notebook.
- The generator framework introduced under generators.
- This implementation.

The plan is to consolidate and move towards a single algorithm rather than splinter into multiple approaches. Can these improvements be incorporated into the generators algorithm?

Reply:
No doubt it is possible. 

The challenge would be the same - to allow for balance across products & categories (if interactions for a particular category are "spread too thin" we can end up with nothing recommended from the category and some items can get forgotten). If more products and categories are in the process of being added to the demo, this may be less of an issue? Note that the old interactions generator also considers gender as a factor which we would need to bring in to the new one (but this was one cause of imbalance). 

Alternatively, we could fold any changes in when the original interaction generator is folded in (it seems this is planned?).